engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java

```diff
@@ -640,6 +479,60 @@ private void loadAllTableDataIntoTheCache() throws KettleException {
     }
   }
 
+  private void putToDefaultCache( Database db, List<Object[]> rows ) {
+    RowMetaInterface returnRowMeta = db.getReturnRowMeta();
+    // Copy the data into 2 parts: key and value...
+    //
+    for ( Object[] row : rows ) {
+      int index = 0;
+      RowMeta keyMeta = new RowMeta();
+      Object[] keyData = new Object[ meta.getStreamKeyField1().length ];
+      for ( int i = 0; i < meta.getStreamKeyField1().length; i++ ) {
+        keyData[ i ] = row[ index ];
+        keyMeta.addValueMeta( returnRowMeta.getValueMeta( index++ ) );
+      }
+      // RowMeta valueMeta = new RowMeta();
+      Object[] valueData = new Object[ data.returnMeta.size() ];
+      for ( int i = 0; i < data.returnMeta.size(); i++ ) {
+        valueData[ i ] = row[ index++ ];
+        // valueMeta.addValueMeta(returnRowMeta.getValueMeta(index++));
+      }
+      // Store the data...
+      //
+      data.cache.storeRowInCache( meta, keyMeta, keyData, valueData );
+      incrementLinesInput();
+    }
+  }
+
+  private void putToReadOnlyCache( Database db, List<Object[]> rows ) {
+    ReadAllCache.Builder cacheBuilder = new ReadAllCache.Builder( data, rows.size() );
+    RowMetaInterface returnRowMeta = db.getReturnRowMeta();
+
+    // all keys have the same row meta,
+    // it is useless to re-create it each time
+    final int keysAmount = meta.getStreamKeyField1().length;
+    RowMeta keyMeta = new RowMeta();
```

Comment:
@akhayrutdinov please consider using RowMeta.clone() as it should now be much faster than adding individual value metas (https://github.com/pentaho/pentaho-kettle/commit/88e0ebbfb0caa67c5f72d19066697d1fa1856793)